src/module.cpp

```diff
@@ -423,61 +421,62 @@ instruction_ref module::move_instructions(instruction_ref src, instruction_ref d
 
 std::vector<instruction_ref>
 module::add_instructions(const std::vector<instruction_ref>& instructions,
-                         std::unordered_map<instruction_ref, instruction_ref> map_ins,
+                         std::unordered_map<instruction_ref, instruction_ref>* map_ins,
                          module::inserter insert)
 {
-    return this->insert_instructions(
-        this->end(), instructions, std::move(map_ins), std::move(insert));
+    return this->insert_instructions(this->end(), instructions, map_ins, std::move(insert));
 }
 
 std::vector<instruction_ref>
 module::add_instructions(const_module_ref m,
-                         std::unordered_map<instruction_ref, instruction_ref> map_ins,
+                         std::unordered_map<instruction_ref, instruction_ref>* map_ins,
                          module::inserter insert)
 {
-    return this->insert_instructions(this->end(), m, std::move(map_ins), std::move(insert));
+    return this->insert_instructions(this->end(), m, map_ins, std::move(insert));
 }
 
 std::vector<instruction_ref>
 module::add_instructions(instruction_ref start,
                          instruction_ref last,
-                         std::unordered_map<instruction_ref, instruction_ref> map_ins,
+                         std::unordered_map<instruction_ref, instruction_ref>* map_ins,
                          module::inserter insert)
 {
-    return this->insert_instructions(
-        this->end(), start, last, std::move(map_ins), std::move(insert));
+    return this->insert_instructions(this->end(), start, last, map_ins, std::move(insert));
 }
 
 std::vector<instruction_ref>
 module::insert_instructions(instruction_ref ins,
                             const std::vector<instruction_ref>& instructions,
-                            std::unordered_map<instruction_ref, instruction_ref> map_ins,
+                            std::unordered_map<instruction_ref, instruction_ref>* map_ins,
                             module::inserter insert)
 {
+    std::unordered_map<instruction_ref, instruction_ref> default_map_ins;
     return insert_generic_instructions(
-        *this, ins, instructions, std::move(map_ins), std::move(insert));
+        *this, ins, instructions, map_ins ? *map_ins : default_map_ins, std::move(insert));
```

Comment:
Wouldn't it be safer to do `!map_ins ? default_map_ins: *map_ins ` here since map_ins can be null. I'm assuming pointer is used vs reference here since the initialization for map_ins can potentially be nullptr by default.

Is there a way to not use nullptr in anyway?

Reply:
> Wouldn't it be safer to do !map_ins ? default_map_ins: *map_ins here since map_ins can be null.

Whats the difference? It seems cleaner to reverse the condition so we dont need the extra `not`.

> Is there a way to not use nullptr in anyway and move these all to references?

No, this is an optional parameter.

Reply:
Ah wait, I need to explictly check for null, so its probably better to invert it then.

Reply:
> Do we not want to just have a default reference to some object instead if its not set just for type safety/needing to worry about null? 

We would still need to check for null though.

> One thing that does irk me is that we're passing around null pointers for map_ins then default setting this after a check, whereas why not just make this a reference to a default empty and pass around const references then? Or is that not possible in this case?

You can't do that with lvalue references. Also as a bonus, with a pointer it will make it explicit from the caller that we will modify the variable.